cmake/CompilerWarnings.cmake

```
@@ -0,0 +1,90 @@
+            /w14928 # illegal copy-initialization; more than one user-defined conversion has been
+                    # implicitly applied
+        )
+
+        set(CLANG_WARNINGS
+            -Wall
+            -Wextra # reasonable and standard
+            -Wshadow # warn the user if a variable declaration shadows one from a parent context
+            -Wnon-virtual-dtor # warn the user if a class with virtual functions has a non-virtual
+                                # destructor. This helps catch hard to track down memory errors
+            -Wold-style-cast # warn for c-style casts
+            -Wcast-align # warn for potential performance problem casts
+            -Wunused # warn on anything being unused
+            -Woverloaded-virtual # warn if you overload (not override) a virtual function
+            -Wpedantic # warn if non-standard C++ is used
+            -Wconversion # warn on type conversions that may lose data
+            -Wsign-conversion # warn on sign conversions
+            -Wnull-dereference # warn if a null dereference is detected
+            -Wdouble-promotion # warn if float is implicit promoted to double
+            -Wformat=2 # warn on security issues around functions that format output (ie printf)
+        )
+        
+        if(WARNINGS_AS_ERRORS)
+            message("warnings as errors")
+            set(CLANG_WARNINGS ${CLANG_WARNINGS} -Werror)
+            set(MSVC_WARNINGS ${MSVC_WARNINGS} /WX)
+        endif()
+
+        if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+            set(CLANG_WARNINGS ${CLANG_WARNINGS} -O0 -g)
+            set(MSVC_WARNINGS ${MSVC_WARNINGS} /Od)
+        elseif(CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
+            set(CLANG_WARNINGS ${CLANG_WARNINGS} -O3 -g)
+            set(MSVC_WARNINGS ${MSVC_WARNINGS} /O2d)
+        elseif(CMAKE_BUILD_TYPE STREQUAL "MinSizeRel")
+            set(CLANG_WARNINGS ${CLANG_WARNINGS} -Os)
+            set(MSVC_WARNINGS ${MSVC_WARNINGS} /O1)
+        else()
+            set(CLANG_WARNINGS ${CLANG_WARNINGS} -O3)
+            set(MSVC_WARNINGS ${MSVC_WARNINGS} /O2)
+        endif()
+
+        set(GCC_WARNINGS
+            ${CLANG_WARNINGS}
+            -Wmisleading-indentation # warn if identation implies blocks where blocks do not exist
+            -Wduplicated-cond # warn if if / else chain has duplicated conditions
+            -Wduplicated-branches # warn if if / else branches have duplicated code
+            -Wlogical-op # warn about logical operations being used where bitwise were probably wanted
+            -Wuseless-cast # warn if you perform a cast to the same type
+        )
+
+        target_compile_options(
+            ${target_name} PRIVATE
+              $<$<OR:$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>>:
+              ${CLANG_WARNINGS}>
+              $<$<CXX_COMPILER_ID:MSVC>:
+              ${MSVC_WARNINGS}>
+              $<$<CXX_COMPILER_ID:GNU>:
+              ${GCC_WARNINGS}>)
```

Can you format in a way it’s easier to read? Like:
```suggestion
            ${target_name} PRIVATE
                 $<$<OR:
                     $<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>>:
                         ${CLANG_WARNINGS}
                     >
                     $<$<CXX_COMPILER_ID:MSVC>:
                         ${MSVC_WARNINGS}
                     >
                     $<$<CXX_COMPILER_ID:GNU>:
                         ${GCC_WARNINGS}
                     >
        )
```

---

Concerning code formatting, I will introduce `cmake-format` tool.